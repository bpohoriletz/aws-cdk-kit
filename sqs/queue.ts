import { Stack, CfnOutput } from 'aws-cdk-lib';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { SqsSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';
import { Queue } from 'aws-cdk-lib/aws-sqs';
import * as cn from '../utils/naming';

export function createOrdersQueue(resourceNamePrefix: string[], topic: Topic, stack: Stack): Queue {
  const resourceName = cn.sqsQueueName(resourceNamePrefix, 'Orders');
  const queue = new Queue(stack, resourceName, {
    fifo: true,
  });
  // Bind the SQS Queue to the SNS Topic.
  const subscription = new SqsSubscription(queue, {
    rawMessageDelivery: true,
  });
  topic.addSubscription(subscription);

  new CfnOutput(stack, 'ordersQueueArn', { value: queue.queueArn });

  return queue;
}
