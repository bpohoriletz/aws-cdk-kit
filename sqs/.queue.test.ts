import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { stub } from '../test/stubs';

import * as func from './queue';

describe('SQS .createOrdersQueue()', () => {
  let stack: cdk.Stack;

  beforeEach(() => {
    stack = new cdk.Stack();
  });

  test('matches snapshot', () => {
    func.createOrdersQueue(['pre', 'fix'], stub(stack, 'sns.Topic'), stack);

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
