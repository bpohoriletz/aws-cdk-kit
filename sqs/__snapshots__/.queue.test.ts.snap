// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQS .createOrdersQueue() matches snapshot 1`] = `
{
  "Outputs": {
    "ordersQueueArn": {
      "Value": {
        "Fn::GetAtt": [
          "PreOrdersQueue4BD5F866",
          "Arn",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "PreOrdersQueue4BD5F866": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "FifoQueue": true,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "PreOrdersQueuePolicyC9DBDAAF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Ref": "TopicID57BEF19B",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "PreOrdersQueue4BD5F866",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "PreOrdersQueue4BD5F866",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "PreOrdersQueueTopicID7AAE870A": {
      "DependsOn": [
        "PreOrdersQueuePolicyC9DBDAAF",
      ],
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "PreOrdersQueue4BD5F866",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "RawMessageDelivery": true,
        "TopicArn": {
          "Ref": "TopicID57BEF19B",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TopicID57BEF19B": {
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
